File: backend/src/Modules/Patients/HMS.Modules.Patients/Domain/Patient.cs
using HMS.Modules.Patients.Contracts;
using HMS.Shared.Kernel;

namespace HMS.Modules.Patients.Domain;

/// <summary>
/// A patient's master/demographic record — the aggregate root for this module. Address is a
/// true 1:1 child (always present, created together with the patient); Allergies and
/// EmergencyContacts are 1:many children ("Add another ..." on the frontend). All three are
/// only ever loaded/saved through this aggregate — see Infrastructure/Repositories/
/// PatientRepository.cs. Encounter/visit data (department, consultant, admission type) is
/// deliberately out of scope for this iteration — see the rebuild plan's Context section.
/// </summary>
internal class Patient : Entity
{
    public string Uhid { get; private set; } = null!;

    public Title Title { get; private set; }
    public string FirstName { get; private set; } = null!;
    public string LastName { get; private set; } = null!;
    public DateOnly DateOfBirth { get; private set; }
    public Gender Gender { get; private set; }
    public BloodGroup BloodGroup { get; private set; }
    public MaritalStatus MaritalStatus { get; private set; }

    public string PrimaryPhone { get; private set; } = null!;
    public string? SecondaryPhone { get; private set; }
    public string? Email { get; private set; }
    public string? Profession { get; private set; }

    public IdProofType? IdProofType { get; private set; }
    public string? IdProofNumber { get; private set; }

    public ModeOfArrivalSource ModeOfArrivalSource { get; private set; }
    public string? ModeOfArrivalChannel { get; private set; }
    public string? ModeOfArrivalSpecify { get; private set; }

    public Address Address { get; private set; } = null!;

    /// <summary>True when this patient was created with placeholder/default values for
    /// required fields (currently: bulk import — see PatientImportCommitBackgroundService),
    /// which are technically valid but not real data (e.g. a synthetic phone number, a
    /// sentinel date of birth). Front-end surfaces a "verify this patient's details" banner
    /// while true, cleared automatically the next time anyone saves an edit through
    /// UpdateAsync — matches how a receptionist naturally re-confirms details with the patient
    /// once they're back in front of them.</summary>
    public bool RequiresDataVerification { get; private set; }

    private readonly List<Allergy> _allergies = [];
    public IReadOnlyCollection<Allergy> Allergies => _allergies.AsReadOnly();

    private readonly List<EmergencyContact> _emergencyContacts = [];
    public IReadOnlyCollection<EmergencyContact> EmergencyContacts => _emergencyContacts.AsReadOnly();

    /// <summary>Age is always derived from <see cref="DateOfBirth"/>, never stored.</summary>
    public int Age => CalculateAge(DateOfBirth, DateOnly.FromDateTime(DateTime.UtcNow));

    // Required by EF Core materialization.
    private Patient()
    {
    }

    private Patient(
        Guid id,
        string uhid,
        Title title,
        string firstName,
        string lastName,
        DateOnly dateOfBirth,
        Gender gender,
        BloodGroup bloodGroup,
        MaritalStatus maritalStatus,
        string primaryPhone,
        string? secondaryPhone,
        string? email,
        string? profession,
        IdProofType? idProofType,
        string? idProofNumber,
        ModeOfArrivalSource modeOfArrivalSource,
        string? modeOfArrivalChannel,
        string? modeOfArrivalSpecify,
        bool requiresDataVerification,
        Guid? createdBy)
        : base(id, createdBy)
    {
        Uhid = uhid;
        Title = title;
        FirstName = firstName;
        LastName = lastName;
        DateOfBirth = dateOfBirth;
        Gender = gender;
        BloodGroup = bloodGroup;
        MaritalStatus = maritalStatus;
        PrimaryPhone = primaryPhone;
        SecondaryPhone = secondaryPhone;
        Email = email;
        Profession = profession;
        IdProofType = idProofType;
        IdProofNumber = idProofNumber;
        ModeOfArrivalSource = modeOfArrivalSource;
        ModeOfArrivalChannel = modeOfArrivalChannel;
        ModeOfArrivalSpecify = modeOfArrivalSpecify;
        RequiresDataVerification = requiresDataVerification;
    }

    public static Patient Create(
        string uhid,
        Title title,
        string firstName,
        string lastName,
        DateOnly dateOfBirth,
        Gender gender,
        BloodGroup bloodGroup,
        MaritalStatus maritalStatus,
        string primaryPhone,
        string? secondaryPhone,
        string? email,
        string? profession,
        IdProofType? idProofType,
        string? idProofNumber,
        ModeOfArrivalSource modeOfArrivalSource,
        string? modeOfArrivalChannel,
        string? modeOfArrivalSpecify,
        Guid? createdBy,
        bool requiresDataVerification = false)
    {
        Guard.AgainstNullOrWhiteSpace(uhid, nameof(uhid));
        Guard.AgainstNullOrWhiteSpace(firstName, nameof(firstName));
        Guard.AgainstNullOrWhiteSpace(lastName, nameof(lastName));
        Guard.AgainstNullOrWhiteSpace(primaryPhone, nameof(primaryPhone));

        return new Patient(
            Guid.CreateVersion7(),
            uhid.Trim(),
            title,
            firstName.Trim(),
            lastName.Trim(),
            dateOfBirth,
            gender,
            bloodGroup,
            maritalStatus,
            primaryPhone.Trim(),
            secondaryPhone?.Trim(),
            email?.Trim().ToLowerInvariant(),
            profession?.Trim(),
            idProofType,
            idProofNumber?.Trim(),
            modeOfArrivalSource,
            modeOfArrivalChannel?.Trim(),
            modeOfArrivalSpecify?.Trim(),
            requiresDataVerification,
            createdBy);
    }

    public void UpdateDemographics(Title title, string firstName, string lastName, DateOnly dateOfBirth, Gender gender, BloodGroup bloodGroup, MaritalStatus maritalStatus, Guid? updatedBy)
    {
        Guard.AgainstNullOrWhiteSpace(firstName, nameof(firstName));
        Guard.AgainstNullOrWhiteSpace(lastName, nameof(lastName));

        Title = title;
        FirstName = firstName.Trim();
        LastName = lastName.Trim();
        DateOfBirth = dateOfBirth;
        Gender = gender;
        BloodGroup = bloodGroup;
        MaritalStatus = maritalStatus;
        MarkUpdated(updatedBy);
    }

    public void UpdateContact(string primaryPhone, string? secondaryPhone, string? email, string? profession, Guid? updatedBy)
    {
        Guard.AgainstNullOrWhiteSpace(primaryPhone, nameof(primaryPhone));

        PrimaryPhone = primaryPhone.Trim();
        SecondaryPhone = secondaryPhone?.Trim();
        Email = email?.Trim().ToLowerInvariant();
        Profession = profession?.Trim();
        MarkUpdated(updatedBy);
    }

    public void UpdateIdProof(IdProofType? idProofType, string? idProofNumber, Guid? updatedBy)
    {
        IdProofType = idProofType;
        IdProofNumber = idProofNumber?.Trim();
        MarkUpdated(updatedBy);
    }

    public void UpdateModeOfArrival(ModeOfArrivalSource source, string? channel, string? specify, Guid? updatedBy)
    {
        ModeOfArrivalSource = source;
        ModeOfArrivalChannel = channel?.Trim();
        ModeOfArrivalSpecify = specify?.Trim();
        MarkUpdated(updatedBy);
    }

    /// <summary>Called once per PatientService.UpdateAsync call, after every field update —
    /// any full save through the Edit Patient screen counts as the receptionist having
    /// re-confirmed the patient's details, regardless of which specific fields they touched.
    /// A no-op (cheap) when the flag was already false.</summary>
    public void ClearDataVerificationFlag(Guid? updatedBy)
    {
        if (!RequiresDataVerification)
        {
            return;
        }

        RequiresDataVerification = false;
        MarkUpdated(updatedBy);
    }

    /// <summary>Set once, at creation — see PatientService.CreateAsync. Address always exists
    /// for a patient, so there's no separate "add" step the way Allergies/EmergencyContacts have.</summary>
    public void SetAddress(Address address) => Address = address;

    public void UpdateAddress(string addressLine1, string? addressLine2, string? addressLine3, Guid stateId, Guid districtId, string pincode, Guid? updatedBy)
    {
        Address.Update(addressLine1, addressLine2, addressLine3, stateId, districtId, pincode);
        MarkUpdated(updatedBy);
    }

    public void AddAllergy(Allergy allergy, Guid? updatedBy)
    {
        _allergies.Add(allergy);
        MarkUpdated(updatedBy);
    }

    /// <returns>false if no allergy with that id exists on this patient.</returns>
    public bool RemoveAllergy(Guid allergyId, Guid? updatedBy)
    {
        var allergy = _allergies.FirstOrDefault(a => a.Id == allergyId);
        if (allergy is null)
        {
            return false;
        }

        _allergies.Remove(allergy);
        MarkUpdated(updatedBy);
        return true;
    }

    public void AddEmergencyContact(EmergencyContact contact, Guid? updatedBy)
    {
        _emergencyContacts.Add(contact);
        MarkUpdated(updatedBy);
    }

    /// <returns>false if no contact with that id exists on this patient, or if it's the
    /// patient's only remaining emergency contact (every patient must have at least one —
    /// same rule CreatePatientRequestValidator enforces at registration time).</returns>
    public bool RemoveEmergencyContact(Guid emergencyContactId, Guid? updatedBy)
    {
        var contact = _emergencyContacts.FirstOrDefault(c => c.Id == emergencyContactId);
        if (contact is null || _emergencyContacts.Count <= 1)
        {
            return false;
        }

        _emergencyContacts.Remove(contact);
        MarkUpdated(updatedBy);
        return true;
    }

    private static int CalculateAge(DateOnly dateOfBirth, DateOnly asOf)
    {
        var age = asOf.Year - dateOfBirth.Year;
        if (dateOfBirth > asOf.AddYears(-age))
        {
            age--;
        }

        return age;
    }
}

File: frontend/shared/dtos/patients/patient.ts
import type {
  AllergySeverity,
  AllergyType,
  BloodGroup,
  Gender,
  IdProofType,
  MaritalStatus,
  ModeOfArrivalSource,
  Relationship,
  Title,
} from '../../enums/patients';

/** Mirrors HMS.Modules.Patients.Contracts.AddressRequest/AddressResponse. */
export interface Address {
  addressLine1: string;
  addressLine2?: string | null;
  addressLine3?: string | null;
  stateId: string;
  districtId: string;
  pincode: string;
}

/** Mirrors HMS.Modules.Patients.Contracts.AllergyResponse (id present) / AllergyRequest (id absent, see AllergyInput below). */
export interface Allergy {
  id: string;
  allergyType: AllergyType;
  specify?: string | null;
  severity: AllergySeverity;
}

export type AllergyInput = Omit<Allergy, 'id'>;

/** Mirrors HMS.Modules.Patients.Contracts.EmergencyContactResponse (id present) / EmergencyContactRequest (id absent, see EmergencyContactInput below). */
export interface EmergencyContact {
  id: string;
  relationship: Relationship;
  name: string;
  phone: string;
}

export type EmergencyContactInput = Omit<EmergencyContact, 'id'>;

/** Mirrors HMS.Modules.Patients.Contracts.PatientResponse. */
export interface Patient {
  id: string;
  uhid: string;

  title: Title;
  firstName: string;
  lastName: string;
  dateOfBirth: string;
  age: number;
  gender: Gender;
  bloodGroup: BloodGroup;
  maritalStatus: MaritalStatus;

  primaryPhone: string;
  secondaryPhone?: string | null;
  email?: string | null;
  profession?: string | null;

  idProofType?: IdProofType | null;
  idProofNumber?: string | null;

  modeOfArrivalSource: ModeOfArrivalSource;
  modeOfArrivalChannel?: string | null;
  modeOfArrivalSpecify?: string | null;

  address: Address;
  allergies: Allergy[];
  emergencyContacts: EmergencyContact[];

  /** Opaque optimistic-concurrency token (the row's Postgres xmin at read time) — echo this
   * back on UpdatePatientRequest.rowVersion so a save against stale data is rejected with a
   * clear conflict instead of silently overwriting someone else's edit. Always present on a
   * real API response (see PatientResponse.RowVersion); the mock store populates a fake one
   * too, so this is never actually undefined at runtime. */
  rowVersion: string;

  createdAt: string;
  updatedAt?: string | null;
}

/** Mirrors HMS.Modules.Patients.Contracts.CreatePatientRequest. */
export interface CreatePatientRequest {
  title: Title;
  firstName: string;
  lastName: string;
  dateOfBirth: string;
  gender: Gender;
  bloodGroup: BloodGroup;
  maritalStatus: MaritalStatus;

  primaryPhone: string;
  secondaryPhone?: string | null;
  email?: string | null;
  profession?: string | null;

  idProofType?: IdProofType | null;
  idProofNumber?: string | null;

  modeOfArrivalSource: ModeOfArrivalSource;
  modeOfArrivalChannel?: string | null;
  modeOfArrivalSpecify?: string | null;

  address: Address;
  allergies: AllergyInput[];
  emergencyContacts: EmergencyContactInput[];
}

/** Mirrors HMS.Modules.Patients.Contracts.UpdatePatientRequest — Allergies/EmergencyContacts
 * have their own add/remove endpoints (AddAllergyRequest/AddEmergencyContactRequest below)
 * rather than being replaced wholesale here. */
export interface UpdatePatientRequest {
  title: Title;
  firstName: string;
  lastName: string;
  dateOfBirth: string;
  gender: Gender;
  bloodGroup: BloodGroup;
  maritalStatus: MaritalStatus;

  primaryPhone: string;
  secondaryPhone?: string | null;
  email?: string | null;
  profession?: string | null;

  idProofType?: IdProofType | null;
  idProofNumber?: string | null;

  modeOfArrivalSource: ModeOfArrivalSource;
  modeOfArrivalChannel?: string | null;
  modeOfArrivalSpecify?: string | null;

  address: Address;

  /** Must be the RowVersion from the Patient this edit was loaded from — see Patient.rowVersion. */
  rowVersion: string;
}

/** Mirrors HMS.Modules.Patients.Contracts.AddAllergyRequest — not wired to any UI yet (Create
 * sends the full Allergies[] up front instead), kept here so the DTO file matches the backend
 * surface completely for the next pass that adds per-row add/remove to the Edit page. */
export type AddAllergyRequest = AllergyInput;

/** Mirrors HMS.Modules.Patients.Contracts.AddEmergencyContactRequest — same "not wired yet" note as AddAllergyRequest. */
export type AddEmergencyContactRequest = EmergencyContactInput;

/** Mirrors HMS.Modules.Patients.Contracts.PatientListQuery. */
export interface PatientListQuery {
  page?: number;
  pageSize?: number;
  sort?: string;
  /** @deprecated superseded by the separate name/age/uhid/phone fields below. */
  search?: string;
  name?: string;
  age?: number;
  uhid?: string;
  phone?: string;
}
